File: benchmarks/cross_language/c/prime_sieve.c
/* Benchmark: Prime counter via trial division up to 100,000 -- C equivalent.
 * Same algorithm as Mapanare list_ops.mn: is_prime checks n < 2, n < 4, n % 2,
 * then odd d starting at 3 up to sqrt(n).
 * Expected: primes = 9592.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <time.h>
#include <sys/resource.h>
#include <sys/time.h>

static bool is_prime(int64_t n) {
    if (n < 2) return false;
    if (n < 4) return true;
    if (n % 2 == 0) return false;
    int64_t d = 3;
    while (d * d <= n) {
        if (n % d == 0) return false;
        d += 2;
    }
    return true;
}

static double tv_to_sec(const struct timeval *tv) {
    return (double)tv->tv_sec + (double)tv->tv_usec / 1e6;
}

static double ts_to_sec(const struct timespec *ts) {
    return (double)ts->tv_sec + (double)ts->tv_nsec / 1e9;
}

int main(void) {
    struct timespec wall0, wall1;
    struct rusage ru0, ru1;

    getrusage(RUSAGE_SELF, &ru0);
    clock_gettime(CLOCK_MONOTONIC, &wall0);

    int64_t count = 0;
    for (int64_t i = 0; i < 100000; i++) {
        if (is_prime(i)) count++;
    }

    clock_gettime(CLOCK_MONOTONIC, &wall1);
    getrusage(RUSAGE_SELF, &ru1);

    double wall = ts_to_sec(&wall1) - ts_to_sec(&wall0);
    double cpu = (tv_to_sec(&ru1.ru_utime) - tv_to_sec(&ru0.ru_utime))
               + (tv_to_sec(&ru1.ru_stime) - tv_to_sec(&ru0.ru_stime));
    double peak_kb = (double)ru1.ru_maxrss;

    printf("primes = %lld\n", (long long)count);
    printf("__BENCH_METRICS__\n");
    printf("wall_time_s=%.6f\n", wall);
    printf("cpu_time_s=%.6f\n", cpu);
    printf("peak_memory_kb=%.1f\n", peak_kb);
    return 0;
}

File: benchmarks/cross_language/go/prime_sieve.go
// Benchmark: Prime counter via trial division up to 100,000 -- Go equivalent.
// Mirrors the Mapanare is_prime (list_ops.mn).
// Expected output: primes = 9592.
package main

import (
	"fmt"
	"runtime"
	"syscall"
	"time"
)

func isPrime(n int64) bool {
	if n < 2 {
		return false
	}
	if n < 4 {
		return true
	}
	if n%2 == 0 {
		return false
	}
	d := int64(3)
	for d*d <= n {
		if n%d == 0 {
			return false
		}
		d += 2
	}
	return true
}

func main() {
	runtime.GC()
	var ruStart, ruEnd syscall.Rusage
	syscall.Getrusage(syscall.RUSAGE_SELF, &ruStart)
	wall0 := time.Now()

	var count int64
	for i := int64(0); i < 100000; i++ {
		if isPrime(i) {
			count++
		}
	}

	wall := time.Since(wall0).Seconds()
	syscall.Getrusage(syscall.RUSAGE_SELF, &ruEnd)
	cpu := timevalSub(ruEnd.Utime, ruStart.Utime) + timevalSub(ruEnd.Stime, ruStart.Stime)
	peakKB := float64(ruEnd.Maxrss)

	fmt.Printf("primes = %d\n", count)
	fmt.Println("__BENCH_METRICS__")
	fmt.Printf("wall_time_s=%.6f\n", wall)
	fmt.Printf("cpu_time_s=%.6f\n", cpu)
	fmt.Printf("peak_memory_kb=%.1f\n", peakKB)
}

func timevalSub(end, start syscall.Timeval) float64 {
	return float64(end.Sec-start.Sec) + float64(end.Usec-start.Usec)/1e6
}
